azure-iot-device/tests/provisioning/internal/test_polling_machine.py

```diff
@@ -0,0 +1,175 @@
+
+
+class TestRegister:
+    def test_register_calls_connect_on_request_response_provider_with_subscribe_action(
+        self, polling_machine, mocker
+    ):
+        mock_init_subscribe_action = mocker.patch(
+            "azure.iot.device.provisioning.internal.polling_machine.SubscribeAction"
+        )
+        mock_subscribe_action = mock_init_subscribe_action.return_value
+
+        polling_machine.on_connect_completed = MagicMock()
+
+        mock_request_response_provider = polling_machine._request_response_provider
+        polling_machine.register()
+
+        mock_request_response_provider.connect.assert_called_once_with(
+            polling_machine.on_connect_completed, post_connect_actions=[mock_subscribe_action]
+        )
+
+    def test_on_subscribe_completed_calls_send_register_request_on_request_response_provider(
+        self, polling_machine, mocker
+    ):
+        mock_init_publish_action = mocker.patch(
+            "azure.iot.device.provisioning.internal.polling_machine.PublishAction"
+        )
+        mock_publish_action = mock_init_publish_action.return_value
+
+        mock_init_query_timer = mocker.patch(
+            "azure.iot.device.provisioning.internal.polling_machine.Timer"
+        )
+        mock_query_timer = mock_init_query_timer.return_value
+        mocker.patch.object(mock_query_timer, "start")
+
+        polling_machine.state = "initializing"
+        mock_request_response_provider = polling_machine._request_response_provider
+        polling_machine.on_subscribe_completed()
+
+        mock_request_response_provider.send_request.assert_called_once_with(mock_publish_action)
+
+
+class TestResponse:
+    @pytest.fixture
+    def fake_registration_state(self):
+        return RegistrationState(
+            fake_device_id,
+            fake_assigned_hub,
+            fake_sub_status,
+            fake_created_dttm,
+            fake_last_update_dttm,
+            fake_etag,
+        )
+
+    @pytest.fixture
+    def registration_result_assigned(self, fake_registration_state):
+        return RegistrationResult(
+            fake_request_id, fake_operation_id, "assigned", fake_registration_state
+        )
+
+    @pytest.fixture
+    def registration_result_assigning(self, fake_registration_state):
+        return RegistrationResult(
+            fake_request_id, fake_operation_id, "assigning", fake_registration_state
+        )
+
+    @pytest.fixture
+    def registration_result_error(self, fake_registration_state):
+        return RegistrationResult(
+            fake_request_id, fake_operation_id, "error", fake_registration_state
+        )
+
+    def test_on_receiving_response_with_status_assigned_calls_callback_of_register_complete(
+        self, polling_machine, registration_result_assigned, mocker
+    ):
+        polling_machine._query_timer = MagicMock()
+        mocker.patch.object(polling_machine._query_timer, "cancel")
+
+        polling_machine.state = "polling"
```

Comment:
>polling_machine.state = "polling" [](start = 8, length = 33)

I don't like forcing a state machine into a particular state unnaturally because it doesn't _really_ put the machine in that state.  By that I mean, in real execution, the machine gets to the "polling" state by going through a set of actions.  Those actions change the state of the object you're testing along the way by setting different attributes and calling different functions.  One example (not the only example) is that the query timer is probably running when you're in the polling state.  In your test, the polling machine's query timer is not running, and that makes your test less accurate because you're simulating something that doesn't really happen in the real world.

Another way to think about it is that you should try as much as possible to have your tests _only_ call public functions on your object and access public attributes.  If you need to do something that is not public in order to test your object, then that is a red flag that your public interface is not complete.  (There are some exceptions to this, but I think they're pretty rare.)